scipy/optimize/minpack.py

```diff
@@ -510,28 +523,39 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         values will all be 1 (if the number of parameters for the function
         can be determined using introspection, otherwise a ValueError
         is raised).
-    sigma : None or M-length sequence, optional
-        If not None, the uncertainties in the ydata array. These are used as
-        weights in the least-squares problem
-        i.e. minimising ``np.sum( ((f(xdata, *popt) - ydata) / sigma)**2 )``
-        If None, the uncertainties are assumed to be 1.
+    sigma : None or M-length sequence or MxM array, optional
+        Determines the uncertainty in `ydata`. If we define residuals as
+        ``e = ydata - f(xdata, *popt)``, then the interpretation of `sigma`
+        depends on its number of dimensions:
+
+            - A 1-d `sigma` should contain values of standard deviations of
+              uncertainties in `ydata`. In this case, the optimized function is
+              ``chisq = sum((e / sigma) ** 2)``.
+
+            - A 2-d `sigma` should contain the covariance matrix of
+              uncertainties in `ydata`. In this case, the optimized function is
+              ``chisq = e.T.dot(inv(sigma)).dot(e)``.
```

Comment:
Minor nitpick: this might be more readable with a matmul syntax, `e.T @ inv(sigma) @ e`.

Also please label it with `.. versionadded:: 0.19.0`